### PR TITLE
Allow using more than 255 Edit Widgets for locallab

### LINKED ISSUE #7148: Can't grab the 38th Selective Editing spot
@Desmis et al., in re: Selective Editing problem.

## Background

1. Unknown if there is a programmatic limitation on # of usable selective editing spots.
2. I am trying to create many spots to create a QR code watermark.
3. Files: [selective-issue.zip](https://github.com/user-attachments/files/16337869/selective-issue.zip)

## Description:

* I have created 37 spots.  Creating any more new or duplicate spots works but I am not able to grab the new spots no matter where they are.

## Reproduce the bug:

1. Open attached png & pp3.
2. Generate an additional new or duplicate spot.
3. Try to grab the spot to move it or resize it.
4. The new spot is not visible to the mouse.

## Expected outcome:

* I would expect to be able to generate more spots to complete the QR code.

##  Version in use:

RawTherapee current dev / macOS 12.7.5 *Monterey*

Screen Shot 2024-07-22 at 11 09 16 AM

Progress so far on the QR thingy:
Screen Shot 2024-07-22 at 11 20 56 AM

### PR DESCRIPTION
**Pull request by Pandagrapher**
@Benitoite 
Fixes issue #7148 

Fix for EditWidgets mode OM_65535
Fix for mode switching between OM_255 and OM_65535

### CODE REVIEW

**Comment by Pandagrapher** [comment]
Hi @Lawrence37
This PR is ready to be merged. With 5.11 release, please tell me (or merge it) when it will be the best moment

Pierre

File: rtgui/editbuffer.cc [hunk 1]
```diff
@@ -108,9 +108,11 @@ int ObjectMOBuffer::getObjectID(const rtengine::Coord& location)
     }
 
     if (objectMode == OM_255) {
-        id = (unsigned char)(*( objectMap->get_data() + location.y * objectMap->get_stride() + location.x ));
+        // In OM_255 mode, size of pixel is 1 Byte (i.e. size of unsigned char)
+        memcpy(&id, ( objectMap->get_data() + location.y * objectMap->get_stride() + sizeof(unsigned char) * location.x ), sizeof(unsigned char));
     } else {
-        id = (unsigned short)(*( objectMap->get_data() + location.y * objectMap->get_stride() + location.x ));
+        // In OM_65535 mode, size of pixel is 2 Bytes (i.e. size on unsigned short)
+        memcpy(&id, ( objectMap->get_data() + location.y * objectMap->get_stride() + sizeof(unsigned short) * location.x ), sizeof(unsigned short));
```
Comment (Lawrence37) [inline rtgui/editbuffer.cc]: If the size of the type is important, isn't it better to use [fixed width types](https://en.cppreference.com/w/cpp/types/integer) like `std::uint8_t` and `std::uint16_t`?

File: rtgui/editwidgets.cc [hunk 2]
```diff
@@ -171,9 +171,20 @@ void Circle::drawToMOChannel (Cairo::RefPtr<Cairo::Context> &cr, unsigned short
...
+            /* In OM_65535 mode, FORMAT_RGB16_565 format is used:
+                - Red is represented with 5 bits (mask = 0xF800, left shift = 11) from 0 to 31
+                - Green is represented with 6 bits (mask = 0x7E0, left shift = 5) from 0 to 63
+                - Blue is represented with 5 bits (mask = 0x1F) from 0 to 31
+            */
+            const double red = (((id + 1) & 0xF800) >> 11) / 31.;
+            const double green = (((id + 1) & 0x7E0) >> 5) / 63.;
+            const double blue = ((id + 1) & 0x1F) / 31.;
+            cr->set_source_rgb (red, green, blue);
         }
```
Comment (Lawrence37) [inline rtgui/editwidgets.cc]: This is a lot of repeated code. Can this be refactored into a function?

**Comment by Lawrence37** [comment]
`release-5.11` has been branched off, so it's safe to continue development on `dev`.

**Comment by Pandagrapher** [comment]
@Lawrence37 With a long delay (sorry), I have took into account your review remarks. If you are ok with, feel free to merge the Pull Request

Pierre

**Comment by Lawrence37** [comment]
Looks good. Thank you @Pandagrapher!

**Pull request merged**